src/components/InputUpload/InputUpload.tsx

```diff
@@ -0,0 +1,109 @@
+"use client";
+
+import React, { useRef, useState } from "react";
+import {
+  FieldValues,
+  Path,
+  UseFormReturn,
+  UseFormRegisterReturn,
+  PathValue,
+} from "react-hook-form";
+import Image from "next/image";
+import { IMAGE_TYPES } from "./InputUpload.constants";
+import * as S from "./InputUpload.styles";
+interface InputUploadProps<T extends FieldValues> {
+  position?: "relative" | "absolute";
+  isPreview?: boolean;
+  register: UseFormRegisterReturn;
+  setValue: UseFormReturn<T>["setValue"];
+  name: Path<T>;
+  children?: React.ReactNode;
+  onPreview?: (preview: string | null, imageSize: { width: number; height: number }) => void;
+}
+
+const InputUpload = <T extends FieldValues>({
+  position = "relative",
+  isPreview = true,
+  register,
+  setValue,
+  name,
+  children,
+  onPreview,
+}: InputUploadProps<T>) => {
+  const inputRef = useRef<null | HTMLInputElement>(null);
+  const [preview, setPreview] = useState<string | null>(null);
+  const [imageSize, setImageSize] = useState({
+    width: 0,
+    height: 0,
+  });
+
+  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.files) {
+      const file = event.target.files[0];
+      const fileType = file.type;
+
+      if (IMAGE_TYPES[fileType]) {
+        const reader = new FileReader();
+        reader.onloadend = () => {
+          const image = document.createElement("img");
+          image.onload = () => {
+            setImageSize({
+              width: image.width,
+              height: image.height,
+            });
+
+            if (onPreview) {
+              onPreview(reader.result as string, { width: image.width, height: image.height });
+            } else {
+              setPreview(reader.result as string);
+            }
+          };
+          image.src = reader.result as string;
+        };
+        reader.readAsDataURL(file);
+      } else {
+        alert("파일 형식이 올바르지 않습니다. 이미지 파일을 업로드해 주세요.");
+        event.target.value = "";
+      }
+    }
+
+    if (event.target.files) {
+      setValue(name, event.target.files as unknown as PathValue<T, Path<T>>);
+    }

```

Comment:
[ P2 ]

요 부분도 `as`를 없앨 수 있지 않을까 싶어요!

```suggestion
    if (typeof event.target.files === "string") {
      setValue(name, event.target.files);
    }
```

Reply:
이 부분 어제 고민 많이 하다가 도저히 안돼서 as 문으로 처리했는데 `event.target.files`의 type 여부만 string 인지 판단하면 됐던 거였군요..?! 🥲

Reply:
근데 제네릭 타입이라 스트링만 허용하신건지 아니면 다른 타입도 넣으려하셨던건지 모르겟어서 ㅠ 이 부분은 의도하신대로 넣으시면 좋을것같아요!

Reply:
생각해 보니까 `event.target.files`의 타입이 `FileList` 타입일 때 setValue에 데이터를 넣어야 해서 해당 조건문은 사용할 수 없겠네요 ㅠㅠ